### ISSUE TITLE
`safari10` option doesn't handle Safari destructuring catch scope bug

### ISSUE THREAD

**Issue by sebakerckhof** [terser-js/terser#117]
**Bug report or Feature request?**
Bug (in safari)

**Version (complete output of `terser -V`)**
3.8.2

**Complete CLI command or `minify()` options used**
`terser -m --safari10 code.js`

**`terser` input**

```js
class Foo {
  foo(ev) {
    try {
      
    } catch ({ message }) {
      
    }
  }
}
```
**`terser` output or error**
Output:
```js
class Foo{foo(s){try{}catch({message:s}){}}}
````
Which results in the error in safari: `Unexpected identifier 's': cannot declare a lexical variable twice: 's'`
Even though I use the `--safari10` option.
Breaks on safari 10, 11, 12 & TP build 65 (didn't test earlier versions)

But I think this is actually a bug not yet known from safari, so I opened a bug report here as well: 
https://bugs.webkit.org/show_bug.cgi?id=189914
Still, I think terser should be able to work around it.

**Expected result**
no error.

**Comment by kzc** [comment]
> Bug report or Feature request?
> Bug

This is a feature request to workaround a Safari bug. The code works correctly on standards compliant ECMAScript engines like FireFox 62, Chrome 67 and Node:
```
$ echo '(new class{f(s){try{throw{m:"PASS"}}catch({m:s}){console.log(s)}}}).f();' | node-v10.4.1 
PASS
```
Thanks for creating the webkit bug ticket. Please ping #117 once the scope of the Safari bug is known so that someone might contribute a PR implementing a `safari10` workaround. Not everyone has access to Safari 10+ to test and it is not clear whether only class methods are subject to this webkit catch clause scope bug, or whether it extends to object literal shorthand methods, arrow functions, regular functions, or only happens with destructuring catch clause arguments.
```
$ echo '({f(s){try{throw{m:"PASS"}}catch({m:s}){console.log(s)}}}).f();' | node
PASS

$ echo '((s)=>{try{throw{m:"PASS"}}catch({m:s}){console.log(s)}})();' | node
PASS

$ echo '(function(s){try{throw{m:"PASS"}}catch({m:s}){console.log(s)}})();' | node
PASS

$ echo '(function(s){try{throw"PASS"}catch(s){console.log(s)}})();' | node
PASS
```

**Comment by kzc** [comment]
Without knowing anything about the extent of this Safari 10+ catch scope bug, it might be possible to work around it with some of the existing `ie8` mangle logic.

Can you try this patch and report your findings?

```patch
diff --git a/lib/scope.js b/lib/scope.js
index 5903cf5..c573a7c 100644
--- a/lib/scope.js
+++ b/lib/scope.js
@@ -301,8 +301,8 @@ AST_Toplevel.DEFMETHOD("figure_out_scope", function(options) {
     });
     self.walk(tw);
 
-    // pass 3: fix up any scoping issue with IE8
-    if (options.ie8) {
+    // pass 3: work around IE8 and Safari catch scope bugs
+    if (options.ie8 || options.safari10) {
         self.walk(new TreeWalker(function(node, descend) {
             if (node instanceof AST_SymbolCatch) {
                 var name = node.name;
@@ -560,7 +560,7 @@ AST_Toplevel.DEFMETHOD("mangle_names", function(options) {
             node.mangled_name = name;
             return true;
         }
-        if (!options.ie8 && node instanceof AST_SymbolCatch) {
+        if (!(options.ie8 || options.safari10) && node instanceof AST_SymbolCatch) {
             to_mangle.push(node.definition());
             return;
         }
diff --git a/test/compress/try-catch.js b/test/compress/try-catch.js
index 9c21aaf..f172b5a 100644
--- a/test/compress/try-catch.js
+++ b/test/compress/try-catch.js
@@ -10,3 +10,86 @@ catch_destructuring_with_sequence: {
     }
     expect_exact: "try{throw{}}catch({xCover=(0,function(){})}){}"
 }
+
+broken_safari_catch_scope: {
+    mangle = {
+        safari10: true,
+    }
+    input: {
+        "AAAAAAAA";
+        "BBBBBBB";
+        new class {
+            f(x) {
+                try {
+                    throw {
+                        m: "PASS"
+                    };
+                } catch ({m: s}) {
+                    console.log(s);
+                }
+            }
+        }().f();
+    }
+    expect: {
+        "AAAAAAAA";
+        "BBBBBBB";
+        new class {
+            f(A) {
+                try {
+                    throw {
+                        m: "PASS"
+                    };
+                } catch ({m: B}) {
+                    console.log(B);
+                }
+            }
+        }().f();
+    }
+    expect_stdout: "PASS"
+    node_version: ">=6"
+}
+
+broken_safari_catch_scope_caveat: {
+    // The `input` of this test reportedly fails on Safari 10+ despite
+    // being valid ECMAScript.
+    // The `expect`ed output of this test will also fail on Safari 10+
+    // despite of the `safari10` `mangle` option being enabled.
+    // This test just exists to prove that both forms will run correctly
+    // on ES spec compliant engines such as V8.
+
+    mangle = {
+        safari10: true,
+    }
+    input: {
+        "AAAAAAAA";
+        "BBBBBBB";
+        new class {
+            f(x) {
+                try {
+                    throw {
+                        m: "PASS"
+                    };
+                } catch ({m: x}) {
+                    console.log(x);
+                }
+            }
+        }().f();
+    }
+    expect: {
+        "AAAAAAAA";
+        "BBBBBBB";
+        new class {
+            f(A) {
+                try {
+                    throw {
+                        m: "PASS"
+                    };
+                } catch ({m: A}) {
+                    console.log(A);
+                }
+            }
+        }().f();
+    }
+    expect_stdout: "PASS"
+    node_version: ">=6"
+}
diff --git a/test/ufuzz.json b/test/ufuzz.json
index ce6b48b..5e46695 100644
--- a/test/ufuzz.json
+++ b/test/ufuzz.json
@@ -37,5 +37,8 @@
             "unsafe_proto": true,
             "unsafe_regexp": true
         }
+    },
+    {
+        "safari10": true
     }
 ]
```
This patch also adds a `safari10` config to reminify tests for increased coverage. It won't prove the reminify tests run correctly on Safari, but will show that the tests continue to work correctly on the various V8 JS engines used by Node.

**Comment by sebakerckhof** [comment]
Thanks for the fast response.

> This is a feature request to workaround a Safari bug. 

You're totally right, however it was only during the writing of this issue that I realized this was not the case the `--safari10` option was meant to cover and this might be a new safari bug.

> it is not clear whether only class methods are subject to this webkit catch clause scope bug, or whether it extends to object literal shorthand methods, arrow functions, regular functions, or only happens with destructuring catch clause arguments.

I did some more testing and expanded the bug report: https://bugs.webkit.org/show_bug.cgi?id=189914
Basically it only happens with destructuring catch clause arguments, both in class functions, normal functions, arrow expressions and object literal shorthand.

> it might be possible to work around it with some of the existing ie8 mangle logic.

Yep, the patch seems to work for all the cases mentioned above!

**Comment by kzc** [comment]
In the meantime you can work around Safari's destructuring catch clause scope bug with the minify options:

```js
{
    safari10: true,
    mangle: {
        ie8: true, // https://github.com/terser-js/terser/issues/117
    }
}
```

Safari appears to be the new IE8:

https://github.com/mishoo/UglifyJS2/issues/1753
https://github.com/mishoo/UglifyJS2/issues/2344
https://github.com/terser-js/terser/issues/117

**Comment by kzc** [comment]
@sebakerckhof Could you please change the issue title to:
```
`safari10` option doesn't handle Safari destructuring catch scope bug
```

**Comment by sebakerckhof** [comment]
Sure, thanks.

**Comment by kzc** [comment]
No sense adding any patch until the WebKit project details the extent of the Safari bug in https://bugs.webkit.org/show_bug.cgi?id=189914 as the ~hack~ workaround may not be comprehensive.

It's a low priority for this project to work around bugs in browsers. Doing so only provides a disincentive for browser makers to fix these bugs. There's also the risk that the Terser workarounds for Safari and IE8 bugs could potentially introduce problems with valid code. At the very least the `safari10` and `ie8` options produce less optimal gzip results.

**Comment by sebakerckhof** [comment]
Yeah, I agree that the full scope should be clear first. And I get the hesitation to work around browser bugs, but many years of web development have made me accept the fact that I have to deal with buggy browsers.
And even if it gets fixed in safari tomorrow, it would still take years before everybody would be on a bug-free safari. But again, I understand your stance on this, I'm just looking if I could be of any help once the full scope is clear.

**Comment by fabiosantoscode** [comment]
Released @kzc's patch as version 3.9.2.
